lib/Conversion/PolynomialToStandard/PolynomialToStandard.cpp

```diff
@@ -802,6 +802,149 @@ func::FuncOp PolynomialToStandard::buildPolynomialModFunc(FunctionType funcType,
+                                        const APInt &cmod) {
+  APInt x = root;
+  APInt primRoot = mulMod(root, root, cmod);
+  SmallVector<APInt> vals(degree / 2);
+  for (unsigned i = 0; i < degree / 2; i++) {
+    vals[i] = x;
+    x = mulMod(x, primRoot, cmod);
+  }
+  return DenseElementsAttr::get(type, vals);
+}
+
+static Value fastNTT(ImplicitLocOpBuilder &b, RankedTensorType &inputType,
+                     Value inputVec, APInt root, APInt cmod) {
+  unsigned degree = inputType.getShape()[0];
+  // Basecase of the algorithm
+  if (degree == 1) {
+    return inputVec;
+  }
+
+  // Type of next iteration
+  auto iterType = inputType.clone({degree / 2});
+
+  // Split the input vector into their odd/even components
+  SmallVector<OpFoldResult> size{b.getIndexAttr(degree / 2)};
+  SmallVector<OpFoldResult> stride{b.getIndexAttr(2)};
+  SmallVector<OpFoldResult> offset{b.getIndexAttr(0)};
+
+  auto evens = b.create<tensor::ExtractSliceOp>(iterType, inputVec, offset,
+                                                size, stride);
+  offset = {b.getIndexAttr(1)};
+  auto odds = b.create<tensor::ExtractSliceOp>(iterType, inputVec, offset, size,
+                                               stride);
+
+  // Recursively iterate with root = root^2 and degree / 2
+  APInt rootSquared = mulMod(root, root, cmod);
+  Value evenNTT = fastNTT(b, iterType, evens.getResult(), rootSquared, cmod);
+  Value oddNTT = fastNTT(b, iterType, odds.getResult(), rootSquared, cmod);
+
+  // Construct a dense cmod to allow for elementwise remainder
+  Value cModVec = b.create<arith::ConstantOp>(
+      iterType,
+      DenseElementsAttr::get(
+          iterType, cmod.sextOrTrunc(iterType.getElementTypeBitWidth())));
+
+  // Implementing Equation 23 from https://ieeexplore.ieee.org/document/10177902
+  //
+  // We have that evenNTT = {A_j}, oddNTT = {B_j} and roots = {root^{2j + 1}}
+  // for all 0 <= j < degree/2
+  auto roots = b.create<arith::ConstantOp>(
+      iterType, constructRoots(iterType, root, degree, cmod));
+
+  // Since roots_j * B_j -> [0, cmod^2) then RemUI will compute the modulus
+  auto rootsB = b.create<arith::MulIOp>(iterType, oddNTT, roots);
+  auto rootsBModded = b.create<arith::RemUIOp>(rootsB, cModVec);
+
+  // Since A + rootsB -> [0, 2 * cmod) then RemUI will compute the modulus
+  auto ctPlus = b.create<arith::AddIOp>(iterType, evenNTT, rootsBModded);
+  auto ctPlusModded = b.create<arith::RemUIOp>(ctPlus, cModVec);
+
+  // Since A - rootsB -> (-cmod, cmod) then we can add cmod such that the
+  // range is shifted to (0, 2 * cmod) and use RemUI to compute the modulus
+  auto ctMinus = b.create<arith::SubIOp>(iterType, evenNTT, rootsBModded);
+  auto ctMinusShifted = b.create<arith::AddIOp>(ctMinus, cModVec);
+  auto ctMinusModded = b.create<arith::RemUIOp>(ctMinusShifted, cModVec);
+
+  // Manually concatenate since tensor::ConcatOp is currently not bufferizable
```

Comment:
Not a problem to do this, but you should be able to get around it by adding this to the pattern set in this pass: https://github.com/llvm/llvm-project/blob/7d4e8c1f3bbfe976f4871c9cf953f76d771b0eda/mlir/include/mlir/Dialect/Tensor/Transforms/Transforms.h#L75

Then create a ConcatOp here and that pattern should automatically convert it to insert_slice ops for you.

Reply:
I think it should work the first way (using applyPartialConversion), but you may need to mark the concat op as an illegal op for it to trigger the pattern, which is a quirk of the dialect conversion framework. If that doesn't work, file a TODO and let's go back to the insert slice style.

Reply:
Nice, making the op illegal did the trick